src/MongrationDotNet/ClientSideDocumentMigration.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using MongoDB.Bson;
+using MongoDB.Driver;
+
+namespace MongrationDotNet
+{
+    public abstract class ClientSideDocumentMigration : Migration
+    {
+        public override string Type { get; } = Constants.ClientSideDocumentMigrationType;
+        public abstract string CollectionName { get; }
+
+        public virtual FilterDefinition<BsonDocument> SearchFilters { get; set; } =
+            FilterDefinition<BsonDocument>.Empty;
+
+        public virtual string UpdateFilterField { get; } = "_id";
+        public virtual int BatchSize { get; } = int.MaxValue;
+
+        public override async Task ExecuteAsync(IMongoDatabase database, ILogger logger)
+        {
+            logger?.LogInformation(LoggingEvents.DocumentMigrationStarted, "Migration started for {collection}",
+                CollectionName);
+
+            var collection = database.GetCollection<BsonDocument>(CollectionName);
+
+            var updated = 0;
+            IEnumerable<BsonDocument> documents;
+            do
+            {
+                documents = (await collection.Find(SearchFilters).ToListAsync()).Skip(updated).Take(BatchSize);
```

Comment:
This still has a concurrency issue where a document will accidentally be missed for migration if another document is deleted from the collection during the client side migration looping. It will be fine for our current use case (we never delete) but add a backlog item in github issues to address it and we can merge this now.